### ISSUE TITLE
Pragma(s) missing from docs

### ISSUE THREAD

**Issue by MasterDuke17** [perl6/doc#752]
It's not in the index, but if you do a site search the word 'pragma' occurs several times. I feel it should have its own section with explanations of the various pragmas, e.g., fatal, v6, variables.

----

New page: https://docs.perl6.org/language/pragmas

**⚠ feel free to edit this list**

- [x] use v6 ([testing#Writing_tests](https://docs.perl6.org/language/testing#Writing_tests))
- [x]  use lib ([modules#Finding_Modules](https://docs.perl6.org/language/modules#Finding_Modules))
- [x]  use variables ([variables#Default_Defined_Variables_Pragma](https://docs.perl6.org/language/variables#Default_Defined_Variables_Pragma))
- [x]  use fatal
- [x]  use soft ([functions#Re-dispatching](https://docs.perl6.org/language/functions#Re-dispatching))
- [x]  use MONKEY-TYPING ([augment](https://docs.perl6.org/syntax/augment))
- [x]  use MONKEY-SEE-NO-EVAL ([EVAL](https://docs.perl6.org/routine/EVAL))
- [x]  use precompilation (see #798)
- [x]  use newline (see #797)
- [x]  use trace (see #800)
- [x]  use strict (see #799)
- [x]  use nqp (see #796)
- [x]  use worries (see #801)
- [x]  use internals (see #809)
- [x]  use invocant (NYI) (see #810)
- [x]  use parameters (NYI) (see #811)
- [x]  use MONKEY-(BRAINS, GUTS, BUSINESS, TRAP, SHINE, WRENCH, BARS) (reserved, but I believe NYI)
- [x]  use experimental
- [x] ↑ and it's experiments (see #812)

**Comment by AlexDaniel** [comment]
Let's create a list of all pragmas?

**⚠ feel free to edit this list**
* use v6 ([testing#Writing_tests](https://docs.perl6.org/language/testing#Writing_tests))
* use lib ([modules#Finding_Modules](https://docs.perl6.org/language/modules#Finding_Modules))
* use variables ([variables#Default_Defined_Variables_Pragma](https://docs.perl6.org/language/variables#Default_Defined_Variables_Pragma))
* use soft ([functions#Re-dispatching](https://docs.perl6.org/language/functions#Re-dispatching))
* use MONKEY-TYPING ([augment](https://docs.perl6.org/syntax/augment))
* use MONKEY-SEE-NO-EVAL ([EVAL](https://docs.perl6.org/routine/EVAL))

**Comment by tbrowder** [comment]
On Sunday, July 24, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> Let's create a list of all pragmas?
>
> *⚠ feel free to edit this list*
>
 This description seems incomplete:

>
>    - use MONKEY-SEE-NO-EVAL (EVAL <https://docs.perl6.org/routine/EVAL>)
>
> The one use example doesn't show or explain the results.

>

**Comment by AlexDaniel** [comment]
@tbrowder sure! This list is for someone who will document these things properly.

By the way, should we have a separate page for all pragmas? Or perhaps we can just improve current docs, make them searchable and that's it?

**Comment by tbrowder** [comment]
On Sunday, July 24, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com> wrote:

> @tbrowder <https://github.com/tbrowder> sure! This list is for someone
> who will document these things properly.
>
> By the way, should we have a separate page for all pragmas? Or perhaps we
> can just improve current docs, make them searchable and that's it?
>
I like the idea of a separate page--easier to see some of the higher-level
organization of our HUGE, but wonderful, language.

>

**Comment by tbrowder** [comment]
On Sunday, July 24, 2016, Tom Browder <tom.browder@gmail.com> wrote:

> On Sunday, July 24, 2016, Aleks-Daniel Jakimenko-Aleksejev <
> notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @tbrowder <https://github.com/tbrowder> sure! This list is for someone
>> who will document these things properly.
>>
>> By the way, should we have a separate page for all pragmas? Or perhaps we
>> can just improve current docs, make them searchable and that's it?
>>
> I like the idea of a separate page--easier to see some of the higher-level
> organization of our HUGE, but wonderful, language.
>

But on the page a link to the definitive source should be good enough.

**Comment by tbrowder** [comment]
On Sunday, July 24, 2016, Tom Browder <tom.browder@gmail.com> wrote:

>
>
> On Sunday, July 24, 2016, Aleks-Daniel Jakimenko-Aleksejev <
> notifications@github.com> wrote:
>
>> Let's create a list of all pragmas?
>>
>> *⚠ feel free to edit this list*
>>
>  This description seems incomplete:
>
>>
>>    - use MONKEY-SEE-NO-EVAL (EVAL <https://docs.perl6.org/routine/EVAL>)
>>
>> The one use example doesn't show or explain the results.
>
>>
And I tried in rakudo REPL and got an error--I have on idea to make it
work--going to irc...

**Comment by gfldex** [comment]
All pragmas are tied to some language feature. Without that context explaining them would be rather hard. So we would end up with a section that contains a short introduction followed by a table with a name and a description and a link that you have to click to understand what that stuff is for. Please apply for Approval of the Redundancy Ministry of Redundancy Approval.

All pragmas need to be in the index (easy to check) and also need to be marked as pramga.
So we need `X<|use $pragmaname (pragma)>` to have them all listen by any possible search term. Whereby we get redundancy already because 'use<WS>' and 'pragma' will both produce a complete list.

I'm not happy with how the index is handled in general but I can't really propose any change before we have a LaTeX randerer to see how we map our index entries to something that is a flat (or not so flat) list, in addition to a searchable list.

**Comment by tbrowder** [comment]
cosed by PR #786

**Comment by AlexDaniel** [comment]
@tbrowder nice!

However, I don't think that this issue is fully resolved. I have added a TODO list and marked this as ``big``.

Also, these pragmas are not searchable yet.

**Comment by tbrowder** [comment]
On Tuesday, August 2, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com> wrote:

> Reopened #752 <https://github.com/perl6/doc/issues/752>.
>
I forgot the X<> but will fix that today.

>

**Comment by tbrowder** [comment]
works now with my last PR ==> merge (PR #793 ); closing this again.

**Comment by AlexDaniel** [comment]
Nice, thanks!

However, some of these pragmas are still not described. I think that it is better to have an issue in this repo, so that people can easily see what is documented and what is not (without the need to grep for TODOs or TBDs), so I will reopen it.

In fact, maybe we should turn some other TBDs and TODOs into issues? Maybe that's not a bad idea?

Don't really want to start a close war on this, but I think that as long as the issue is not fully resolved we should not be closing the issues.

**Comment by tbrowder** [comment]
On Thursday, August 4, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com> wrote:

> Nice, thanks!
>
> However, some of these pragmas are still not described. I think that it is
> better to have an issue in this repo, so that people can easily see what is
> documented and what is not (without the need to grep for TODOs or TBDs), so
> I will reopen it.
>
> In fact, maybe we should turn some other TBDs and TODOs into issues? Maybe
> that's not a bad idea?
>
Agree, putting the issues in plain site in bite-sized pieces brings more
attention and maybe encourages stomping them out.

> Don't really want to start a close war on this, but I think that as long
> as the issue is not fully resolved we should not be closing the issues.
>
No argument here!

>

**Comment by tbrowder** [comment]
On Thursday, August 4, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com> wrote
...

> In fact, maybe we should turn some other TBDs and TODOs into issues? Maybe
> that's not a bad idea?
>
So done.

>

**Comment by AlexDaniel** [comment]
Hm. Creating a separate issue for each item is OK too. It looks a bit weird to have so many open issues on a very similar topic, but it is probably much easier to work on the issues this way. However, when I said “other TBDs and TODOs” I meant non-pragma stuff, meaning that we should probably try to keep our TODOs on github instead of having some issues on github and other in the source. Having one ｢big｣ issue is sometimes OK, at least, I'm not willing to split every ｢big｣ issue into separate issues. But yeah, I should not be changing topics that fast :)

Having so many MONKEY-* issues around, does anybody have any idea why these pragmas were reserved? For example, when Larry reserved ``MONKEY-SHINE``, what was he thinking? Why “shine”? And if nobody knows, what could it mean potentially?

**Comment by AlexDaniel** [comment]
OK, given that there is now a separate issue for each pragma we can close this.

#796 #797 #798 #799 #800 #801 #802 #803 #804 #805 #806 #807 #808 #809 #810 #811 #812

**Comment by tbrowder** [comment]
On Friday, August 5, 2016, Aleks-Daniel Jakimenko-Aleksejev <
notifications@github.com> wrote:

> Hm. Creating a separate issue for each item is OK too. It looks a bit
> weird to have so many open issues on a very similar topic, but it is
> probably much easier to work on the issues this way. However, when I said
> “other TBDs and TODOs” I meant non-pragma stuff, meaning that we should
> probably try to keep our TODOs on github instead of having some issues on
> github and other in the source. Having one ｢big｣ issue is sometimes OK, at
> least, I'm not willing to split every ｢big｣ issue into separate issues. But
> yeah, I should not be changing topics that fast :)
>

Maybe the pragma NYIs should be issues on rt.perl?

>

**Comment by AlexDaniel** [comment]
@tbrowder for “invocant” and “parameters”? Yes.
```
perl6 -e 'use parameters'
===SORRY!=== Error while compiling -e
use parameters not yet implemented. Sorry. 
at -e:1
------> use parameters⏏<EOL>


alex@librebook:~$ perl6 -e 'use invocant'
===SORRY!=== Error while compiling -e
use invocant not yet implemented. Sorry. 
at -e:1
------> use invocant⏏<EOL>
```

**Comment by tbrowder** [comment]
On Fri, Aug 5, 2016 at 3:01 PM, Aleks-Daniel Jakimenko-Aleksejev
<notifications@github.com> wrote:
>
> @tbrowder for “invocant” and “parameters”? Yes.
...

Also "internals":

perl6 -e 'use internals'
===SORRY!=== Error while compiling -e
use internals not yet implemented. Sorry.

But these give no response (e.g., implemented but not documented?):

perl6 -e 'use MONKEY-WRENCH'
perl6 -e 'use MONKEY-TYPING'
perl6 -e 'use MONKEY-TRAP'
perl6 -e 'use MONKEY-SHINE'
perl6 -e 'use MONKEY-BUSINESS'
perl6 -e 'use MONKEY-GUTS'
perl6 -e 'use MONKEY-BRAINS'
perl6 -e 'use MONKEY-BARS'

**Comment by AlexDaniel** [comment]
IRC log: http://irclog.perlgeek.de/perl6/2016-08-05#i_12975389

**Comment by tbrowder** [comment]
On Fri, Aug 5, 2016 at 5:12 PM, Aleks-Daniel Jakimenko-Aleksejev
<notifications@github.com> wrote:
>
> IRC log: http://irclog.perlgeek.de/perl6/2016-08-05#i_12975389

So, @Aleks-Daniel, what shall be done with the Pragmas page and the
doc issues and RT issues?

**Comment by AlexDaniel** [comment]
Alright, so after thinking about it:
* Nothing should be done with the Pragmas page, except for filling in missing stuff. That is, we need actual explanations for every “TBD”. There are now issues for each missing description, so everything is fine.
* I think that it is right to have an RT ticket for every NYI. So three tickets should be filed: invocant, parameters and internals (unless the tickets are already there, I was unable to find anything).
* I am not sure what has to be done with doc issues. I don't like NOTSPECCED issues floating there like this (especially that amount), so I have the following idea: we should tag stuff as NOTSPECCED, but if nothing can be done now, we can close the issue. Once something is implemented, we can reopen it.

In other words: if there is nothing we can do, then we consider the issue solved at this particular moment in time. This would clear the list of open issues, and the NOTSPECCED label will make it possible to check the list periodically to make sure that nothing falls through the cracks.

Does it sound sane?

**Comment by AlexDaniel** [comment]
Yes, it does.

TL;DR: we close issue if the thing is not in roast, but we mark it as NOTSPECCED so that we can easily check it again later.

**Comment by AlexDaniel** [comment]
Also, remove NOTSPECCED tag if it turns out that this thing will never need documentation.

**Comment by JJ** [comment]
Unless you're adding a new pragma to the list above, I think all of them have been added, right?
